docs/design/generics/details.md

```diff
@@ -2794,6 +2813,29 @@ constraint ContainerIsSlice {
 Note that using the `constraint` approach we can name these constraints using
 `Self` instead of `.Self`, since they refer to the same type.
 
+The `.Self` construct follows these rules:
+
+-   `X :!` introduces `.Self:! Type`, where references to `.Self` are resolved
+    to `X`. This allows you to use `.Self` as an interface parameter as in
+    `X:! I(.Self)`.
+-   `A where` introduces `.Self:! A` and `.Foo` for each member `Foo` of `A`
+-   It's an error to reference `.Self` if it refers to more than one different
+    thing or isn't a type.
+-   You get the innermost, most-specific type for `.Self` if it is introduced
+    twice in a scope. By the previous rule, it is only legal if they all refer
+    to the same generic parameter.
+
+So in `X:! A where ...`, `.Self` is introduced twice, after the `:!` and the
+`where`. This is allowed since both times it means `X`. After the `:!`, `.Self`
+has the type `Type`, which gets refined to `A` after the `where`. In contrast,
+it is an error if `.Self` could mean two different things, as in:
+
+```
+// ❌ Illegal: `.Self` could mean `T` or `T.A`.
+fn F[T:! InterfaceA where .A is
+           (InterfaceB where .B == .Self)](x: T);
```

Comment:
Getting a bit into the fine details here, what happens for:
```
fn F[T:! InterfaceA where .Self is 
             (InterfaceB where .B == .Self)](x: T);
```
? The second `.Self` arguably only means one thing here -- it resolves to `T` -- but that fact is somewhat obscured via the indirection through the first `.Self`.

Reply:
As written, I agree that would be allowed. Do you think that should change? We could instead have a rule about `is` expression in `where` clauses. What happens when we greatly expand the kinds of expressions in `where` clauses for templates?

I'm mostly looking for a rule with a concise and easy-to-remember statement.

Reply:
I think we also need to think about implementability, for example for:
```
fn F[T:! Identity(InterfaceA where .T == .Self)]();
```
I'd prefer a rule that can be checked immediately and locally. I think the current rule would require an instance of `.Self` to track all the possible results until we reach some enclosing context where we can resolve them and see if they're the same. Maybe we can special-case a `where` expression as the top-level expression on the right of a `:!` and treat all other cases as being immediately ambiguous?